src/bin/pg_dump/dump_babel_utils.c

```diff
@@ -541,3 +612,326 @@ updateExtConfigArray(Archive *fout, char ***extconfigarray, int nconfigitems)
+
+		schema_name = pg_strdup(PQgetvalue(res, i, PQfnumber(res, "nspname")));
+		simple_string_list_append(schema_include_patterns, schema_name);
+		pfree(schema_name);
+	}
+
+	PQclear(res);
+	destroyPQExpBuffer(query);
+}
+
+/*
+ * getBabelfishDependencies:
+ * Sets required dependencies for babelfish objects.
+ */
+void
+getBabelfishDependencies(Archive *fout)
+{
+	PQExpBuffer		query;
+	PGresult		*res;
+	TableInfo		*sysdb_table;
+	TableInfo		*namespace_ext_table;
+	DumpableObject		*dobj;
+	DumpableObject		*refdobj;
+
+	if (!isBabelfishDatabase(fout) || fout->dopt->binary_upgrade)
+		return;
+
+	query = createPQExpBuffer();
+	/* get oids of sys.babelfish_sysdatabases and sys.babelfish_namespace_ext tables */
+	appendPQExpBufferStr(query,
+						 "SELECT oid "
+						 "FROM pg_class "
+						 "WHERE relname in ('babelfish_sysdatabases', 'babelfish_namespace_ext') "
+						 "AND relnamespace = 'sys'::regnamespace "
+						 "ORDER BY relname;");
+	res = ExecuteSqlQuery(fout, query->data, PGRES_TUPLES_OK);
+	
+	Assert(PQntuples(res) == 2);
+	namespace_ext_table = findTableByOid(atooid(PQgetvalue(res, 0, 0)));
+	sysdb_table = findTableByOid(atooid(PQgetvalue(res, 1, 0)));
+	Assert(sysdb_table != NULL && namespace_ext_table != NULL);
+	dobj = (DumpableObject *) namespace_ext_table->dataObj;
+	refdobj = (DumpableObject *) sysdb_table->dataObj;
+	/*
+	 * Make babelfish_namespace_ext table dependent babelfish_sysdatabases
+	 * table so that we dump babelfish_sysdatabases's data before babelfish_namespace_ext.
+	 * This is needed to generate and handle new "dbid" during logical database restore.
+	 */
+	addObjectDependency(dobj, refdobj->dumpId);
+
+	PQclear(res);
+	destroyPQExpBuffer(query);
+}
+
+/*
+ * getCursorForBbfCatalogTableData:
+ * Prepare custom cursor for all Babelfish catalog tables to selectively dump the data
+ * corresponding to specified logical database.
+ */
+void
+getCursorForBbfCatalogTableData(Archive *fout, TableInfo *tbinfo, PQExpBuffer buf, int *nfields)
+{
+	int 	i;
+	bool	is_builtin_db = false;
+
+	if (!isBabelfishDatabase(fout) || bbf_db_name == NULL)
+		return;
+
+	if (tbinfo->dobj.namespace == NULL ||
+		strcmp(tbinfo->dobj.namespace->dobj.name, "sys") != 0 ||
+		!(strcmp(tbinfo->dobj.name, "babelfish_sysdatabases") == 0 ||
+		 strcmp(tbinfo->dobj.name, "babelfish_namespace_ext") == 0 ||
+		 strcmp(tbinfo->dobj.name, "babelfish_function_ext") == 0 ||
+		 strcmp(tbinfo->dobj.name, "babelfish_authid_user_ext") == 0 ||
+		 strcmp(tbinfo->dobj.name, "babelfish_view_def") == 0 ||
+		 strcmp(tbinfo->dobj.name, "babelfish_domain_mapping") == 0))
+		return;
+
+	is_builtin_db = (pg_strcasecmp(bbf_db_name, "master") == 0 ||
+			pg_strcasecmp(bbf_db_name, "tempdb") == 0 ||
+			pg_strcasecmp(bbf_db_name, "msdb") == 0)
+			? true : false;
+
+	resetPQExpBuffer(buf);
+	appendPQExpBufferStr(buf, "DECLARE _pg_dump_cursor CURSOR FOR SELECT ");
+	*nfields = 0;
+	for (i = 0; i < tbinfo->numatts; i++)
+	{
+		if (tbinfo->attisdropped[i])
```

Comment:
nit; for future safety, let's add the corresponding logic or throw errors for generated columns, e.g.,
```
		if (tbinfo->attgenerated[i] && dopt->column_inserts)
			continue;
		if (tbinfo->attgenerated[i])
			appendPQExpBufferStr(q, "NULL");
		else
			appendPQExpBufferStr(q, fmtId(tbinfo->attnames[i]));
		attgenerated[nfields] = tbinfo->attgenerated[i];
```

Reply:
Done!